src/main/java/tc/oc/pgm/discord/DiscordClient.java

```diff
@@ -0,0 +1,37 @@
+package tc.oc.pgm.discord;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import javax.security.auth.login.LoginException;
+import net.dv8tion.jda.api.JDA;
+import net.dv8tion.jda.api.JDABuilder;
+import net.dv8tion.jda.api.entities.User;
+import tc.oc.pgm.Config;
+
+public class DiscordClient {
+
+  private static JDA jda;
+  public static final Cache<UUID, String> TOKENS_CACHE =
+      CacheBuilder.newBuilder().expireAfterWrite(15, TimeUnit.MINUTES).build();
```

Comment:
So this will get a little more complex, but this needs to be handled in `Datastore`.

```java
// Already exists in tc.oc.api
public interface Datastore {

  // Get the discord user
  // If not registered it will still return a DiscordUser, but #getUsername() should be null
  DiscordUser getDiscord(UUID uuid);
}

// Will need to create this in tc.oc.api.discord
public interface DiscordUser {
   // Gets the Discord username of the user, or null if not registered
   @Nullable
   String getUsername();

   // Sets the Discord username
   void setUsername(String username);

   // Gets verification token for a user, cached or un-cached
   String getToken();
}
```

Then you will need to implement these methods in `DatastoreImpl`. Take a look at how `Username` and `Settings` are implemented, which should help you. We need to store the Discord usernames (and verification tokens!) in SQLite.

Reply:
We certainly need to store confirmed discords in the datastore, but i think the tokens are expiring and can safely be stored in memory, they're short-lived, and can be volatile (not expected to survive restarts) wich is fine.

Reply:
> So this will get a little more complex, but this needs to be handled in `Datastore`.
> 
> ```java
> // Already exists in tc.oc.api
> public interface Datastore {
> 
>   // Get the discord user
>   // If not registered it will still return a DiscordUser, but #getUsername() should be null
>   DiscordUser getDiscord(UUID uuid);
> }
> 
> // Will need to create this in tc.oc.api.discord
> public interface DiscordUser {
>    // Gets the Discord username of the user, or null if not registered
>    @Nullable
>    String getUsername();
> 
>    // Sets the Discord username
>    void setUsername(String username);
> 
>    // Gets verification token for a user, cached or un-cached
>    String getToken();
> }
> ```
> 
> Then you will need to implement these methods in `DatastoreImpl`. Take a look at how `Username` and `Settings` are implemented, which should help you. We need to store the Discord usernames (and verification tokens!) in SQLite.

Why would we need to store tokens?

Reply:
You don't need to store the tokens in `SQL`, but I would still use the interface.